### ISSUE TITLE
Bug in restrict / modax functions on AxisArray

### ISSUE THREAD

**Issue by phlavenk** [JuliaImages/Images.jl#657]
The following AxisArrays
```julia
using AxisArrays, Images
img = testimage("cameraman")
img = [img img; img img; img img]
img = img[1:884, 1:1024]
sy, sx = size(img)
imga = AxisArray(img, Axis{:y}(linspace(0.0013497582515,0.0011341817485000002,sy)), Axis{:x}(linspace(1, 2, sx)));
imgb = AxisArray(img[1:5, 1:10], Axis{:y}(linspace(1,5,5)), Axis{:x}(linspace(1, 10, 10)));
```
have problems in the `Images.restrict` function. This makes problems with Jupyter as thi is called inside `Images.downsize_for_thumbnail` function whenever I want to see the image.
* First failing mode:
```julia
julia> Images.restrict(imga, (1,2))                                                                    
ERROR: ArgumentError: the length of each axis must match the corresponding size of data                
 in AxisArrays.AxisArray{T,N,D,Ax}(::Array{ColorTypes.Gray{Float32},2}, ::Tuple{AxisArrays.Axis{:y,Floa
tRange{Float64}},AxisArrays.Axis{:x,FloatRange{Float64}}}) at C:\JuliaPro-0.5.1.1\pkgs-0.5.1.1\v0.5\Axi
sArrays\src\core.jl:206                                                                                
 in restrict(::AxisArrays.AxisArray{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2,Array{ColorTyp
es.Gray{FixedPointNumbers.Normed{UInt8,8}},2},Tuple{AxisArrays.Axis{:y,LinSpace{Float64}},AxisArrays.Ax
is{:x,LinSpace{Float64}}}}, ::Tuple{Int64,Int64}) at C:\JuliaPro-0.5.1.1\pkgs-0.5.1.1\v0.5\Images\src\a
lgorithms.jl:505                                                                                       
```                                                                                                       
Here, the calculation of 
```julia
ay = Axis{:y}(linspace(0.0013497582515,0.0011341817485000002,884))
ay1 = Images.modax(ay)
length(ay1) 
```
yelds FloatRange of wrong length 442 instead of 443 caused by ronunding errors in the Linspace -> FloatRange calculation (that could be verified with other numbers, e.g. `ay = Axis{:y}(linspace(0.0013,0.0011,884))` 
* Second failing mode is in case the AxisArray's axis is based on Linspace and the length is even (case with imgb)
```julia
julia> Images.restrict(imgb, (1,2))
ERROR: MethodError: Cannot `convert` an object of type LinSpace{Float64} to an object of type FloatRange{Float64}
This may have arisen from a call to the constructor FloatRange{Float64}(...),
since type constructors fall back to convert methods.
 in modax(::AxisArrays.Axis{:y,LinSpace{Float64}}) at C:\JuliaPro-0.5.1.1\pkgs-0.5.1.1\v0.5\Images\src\algorithms.jl:524
 in modax at C:\JuliaPro-0.5.1.1\pkgs-0.5.1.1\v0.5\Images\src\algorithms.jl:527 [inlined]
 in map(::Images.#modax, ::Tuple{AxisArrays.Axis{:y,LinSpace{Float64}},AxisArrays.Axis{:x,LinSpace{Float64}}}, ::Tuple{Bool,Bool}) at .\tuple.jl:110
 in restrict(::AxisArrays.AxisArray{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2,Array{ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}},2},Tuple{AxisArrays.Axis{:y,LinSpace{Float64}},AxisArrays.Axis{:x,LinSpace{Float64}}}}, ::Tuple{Int64,Int64}) at C:\JuliaPro-0.5.1.1\pkgs-0.5.1.1\v0.5\Images\src\algorithms.jl:505
```
caused by bug in 
```julia
julia> ay = Axis{:y}(linspace(1,5,5));                                                
julia> ay1 = Images.modax(ay)                                                                          
ERROR: MethodError: Cannot `convert` an object of type LinSpace{Float64} to an object of type FloatRang
e{Float64}                                                                                             
This may have arisen from a call to the constructor FloatRange{Float64}(...),                          
since type constructors fall back to convert methods.                                                  
 in modax(::AxisArrays.Axis{:y,LinSpace{Float64}}) at C:\JuliaPro-0.5.1.1\pkgs-0.5.1.1\v0.5\Images\src\
algorithms.jl:524                                                                                      
```
I'm using JuliaPro 0.5.1 with updated packages
Julia Version 0.5.1                                                
Commit 6445c82 (2017-03-05 13:25 UTC)                              
Platform Info:                                                     
  OS: Windows (x86_64-w64-mingw32)                                 
  CPU: Intel(R) Core(TM) i5-4200M CPU @ 2.50GHz                    
  WORD_SIZE: 64                                                    
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell) 
  LAPACK: libopenblas64_                                           
  LIBM: libopenlibm                                                
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)                            

 - Images                        0.10.1
 - AxisArrays                    0.1.4

I'm aware Tim Holy did huge refactor of ranges/linspaces etc in Julia 0.6 and I don't know if the bugs are solved by 0.6. Unfortunatelly I'm not planning to switch to 0.6 before availability of JuliaPro with working Gallium. 
Befor attempting to write a PR i'd like to clarify for my self why is the `Images.modax` function written the way it is written 
* why the downsizing uses the rule `length(downsized) = floor(length(orig)/2)+1 
* is there a more generic way to "decimate" an arbitrary `Range` object (or even better an arbitrary `AbstractArray` using traits)?
Thank you for the help

**Comment by timholy** [comment]
Images 0.11.1 fixed #655 which seems related (as in, it probably fixes this bug). For 0.5 you could manually backport the fix? (See #656.)

**Comment by timholy** [comment]
All of these work on julia 0.6 and recent package releases.
